pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,437 @@
+		}
+		return requests
+	})
+}
+
+const clusterAndNamespaceDelimiter = "_"
+
+func decodeRequest(req reconcile.Request) (string, types.NamespacedName, error) {
+	clusterAndNamespace := strings.Split(req.Namespace, "_")
+	if n := len(clusterAndNamespace); n != 2 {
+		return "", types.NamespacedName{}, fmt.Errorf("didn't get two but %d segments when trying to extract cluster and namespace", n)
+	}
+	return clusterAndNamespace[0], types.NamespacedName{Namespace: clusterAndNamespace[1], Name: req.Name}, nil
+}
+
+type reconciler struct {
+	ctx                 context.Context
+	log                 *logrus.Entry
+	registryClients     map[string]ctrlruntimeclient.Client
+	pullSecretGetter    func() []byte
+	forbiddenRegistries sets.String
+	dryRun              bool
+}
+
+func (r *reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	log.Info("Starting reconciliation")
+	err := r.reconcile(req, log)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(req reconcile.Request, log *logrus.Entry) error {
+	cluster, decoded, err := decodeRequest(req)
+	if err != nil {
+		return fmt.Errorf("failed to decode request %s: %w", req, err)
+	}
+
+	// Propagate the cluster field back up
+	*log = *log.WithField("cluster", cluster)
+
+	isTags := map[string]*imagev1.ImageStreamTag{}
+	for ctx, client := range r.registryClients {
+		imageStreamTag := &imagev1.ImageStreamTag{}
+		if err := client.Get(r.ctx, decoded, imageStreamTag); err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Debug("Source imageStreamTag not found")
+				isTags[ctx] = nil
+				return nil
+			}
+			return fmt.Errorf("failed to get imageStreamTag %s from registry cluster: %w", decoded.String(), err)
+		}
+		isTags[ctx] = imageStreamTag
+	}
+
+	srcContext, err := findNewest(isTags)
+	if err != nil {
+		return fmt.Errorf("failed to find the new isTag from %v: %w", isTags, err)
+	}
+	sourceImageStreamTag := isTags[srcContext]
+
+	imageStreamNameAndTag := strings.Split(decoded.Name, ":")
+	if n := len(imageStreamNameAndTag); n != 2 {
+		return fmt.Errorf("when splitting imagestreamtagname %s by : expected two results, got %d", decoded.Name, n)
+	}
+	imageStreamName, imageTag := imageStreamNameAndTag[0], imageStreamNameAndTag[1]
+	isName := types.NamespacedName{Namespace: decoded.Namespace, Name: imageStreamName}
+	sourceImageStream := &imagev1.ImageStream{}
+	if err := r.registryClients[srcContext].Get(r.ctx, isName, sourceImageStream); err != nil {
+		return fmt.Errorf("failed to get imageStream %s from registry cluster: %w", isName.String(), err)
+	}
+
+	*log = *log.WithField("docker_image_reference", sourceImageStreamTag.Image.DockerImageReference)
+	if isImportForbidden(sourceImageStreamTag.Image.DockerImageReference, r.forbiddenRegistries) {
```

Comment:
We don't need this check, hence we also don't need to get the `sourceImageStream`

Reply:
Removed `client.get`.
I think we still need `ensureImageStream(sourceImageStream, client, log)`: Can we create `imageStreamImport` without the `IS`?